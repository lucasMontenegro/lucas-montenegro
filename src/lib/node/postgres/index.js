import Bluebird from "bluebird"
import { errorLogger, addIgnoredError } from "./errorFiltering"
import extendRepo from "./extendRepo"
import pgPromise from "pg-promise"
import globals from "lib/utils/globals"
import monitor from "pg-monitor"
import filterMonitorLog from "./filterMonitorLog"
import isProduction from "lib/utils/isProduction"
const initOptions = {
  promiseLib: Bluebird,
  error: errorLogger,
  extend: extendRepo,
}
export { addIgnoredError }
export const pgp = pgPromise(initOptions)
export const db = pgp({
  host: globals.process.env.PG_HOST,
  database: globals.process.env.PG_DATABASE,
  user: globals.process.env.PG_USER,
  password: globals.process.env.PG_PASSWORD,
})
monitor.setTheme(`matrix`)
monitor.setLog(filterMonitorLog)
monitor.attach(initOptions, (
  isProduction() ? [] :
  [`connect`, `disconnect`, `task`, `transact`, `query`]
))
export function closeDb () {
  pgp.end()
}