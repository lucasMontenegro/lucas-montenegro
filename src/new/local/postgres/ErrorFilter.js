import monitor from "pg-monitor"
export default function ErrorFilter () {
  const matchers = []
  return {
    ignore (fn) {
      matchers.push(fn)
    },
    logger (err, e) {
      // usage:
      // pg.ignoreDbError(function matchUniqueUsernameError (err, e) {
      //   return (
      //     typeof err === `object`
      //     && err.code == `23505`
      //     && err.table == `usernames`
      //     && err.constraint == `usernames_pkey`
      //   )
      // })
      // pg.ignoreDbError === logger
      if (!matchers.find(fn => fn(err, e))) {
        monitor.error(err, e)
      }
    },
  }
}