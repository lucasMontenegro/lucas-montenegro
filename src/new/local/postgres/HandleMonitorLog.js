const { default: saveMonitorError } = require("new/local/postgres/saveMonitorError")
const { default: isProduction } = require("new/local/utils/isProduction")
exports.default = function HandleMonitorLog () {
  const production = isProduction()
  return function handleMonitorLog (msg, info) {
    if (info.event === `error`) {
      production || saveMonitorError(msg, info)
    } else if (production) {
      info.display = false
    }
  }
}