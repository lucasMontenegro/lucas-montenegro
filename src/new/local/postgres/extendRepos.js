// Extend "db" object, i.e.:
// const { default: Repo } = require("new/local/Repo")
exports.default = function extendRepos (db, dc) {
  // cont.:
  // db.repo = new Repo(db, dc)
}