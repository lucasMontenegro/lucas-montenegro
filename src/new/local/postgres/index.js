const { default: startDb } = require("new/local/postgres/startDb")
const { db, pgp, ignoreError } = startDb()
exports.pgp = pgp
exports.db = db
exports.ignoreError = ignoreError
exports.closeDb = function closeDb () {
  db.$config.pgp.end()
}