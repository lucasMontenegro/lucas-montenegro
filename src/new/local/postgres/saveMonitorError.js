const os = require("os")
const fs = require("fs")
const path = require("path")
const { default: simpleErrorCallback } = require("new/local/utils/simpleErrorCallback")
const { default: globals } = require("new/local/utils/globals")
const logFile = path.resolve(globals.process.env.NODE_PATH, `../pg-errors.log`)
exports.default = function saveMonitorError (msg, info) {
  let logText = os.EOL + msg
  if (info.time) {
    logText = os.EOL + logText
  }
  fs.appendFile(logFile, logText, simpleErrorCallback)
}