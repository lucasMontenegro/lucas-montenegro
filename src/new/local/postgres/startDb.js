const pgPromise = require("pg-promise")
const monitor = require("pg-monitor")
const Bluebird = require("bluebird")
const { default: HandleMonitorLog } = require("new/local/postgres/HandleMonitorLog")
const { default: ErrorFilter } = require("new/local/postgres/ErrorFilter")
const { default: extendRepos } = require("new/local/postgres/extendRepos")
const { default: globals } = require("new/local/utils/globals")
const { default: isProduction } = require("new/local/utils/isProduction")
exports.default = function startDb () {
  const errorFilter = ErrorFilter()
  const initOptions = {
    promiseLib: Bluebird,
    error: errorFilter.logger,
    extend: extendRepos,
  }
  const pgp = pgPromise(initOptions)
  const db = pgp({
    host: globals.process.env.PG_HOST,
    database: globals.process.env.PG_DATABASE,
    user: globals.process.env.PG_USER,
    password: globals.process.env.PG_PASSWORD,
  })
  monitor.setTheme(`matrix`)
  monitor.setLog(HandleMonitorLog())
  monitor.attach(initOptions, (
    isProduction() ? [] :
    [`connect`, `disconnect`, `task`, `transact`, `query`]
  ))
  return { db, pgp, ignoreError: errorFilter.ignore }
}