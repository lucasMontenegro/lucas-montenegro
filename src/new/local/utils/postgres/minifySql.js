import pgMinify from "pg-minify"
export default function minifySql (str) {
  return pgMinify(str, { compress: true, removeAll: true })
}