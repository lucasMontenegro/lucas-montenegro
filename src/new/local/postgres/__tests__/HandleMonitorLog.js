const mockFn = jest.fn()
jest.mock(`new/local/postgres/saveMonitorError`, () => ({}))
require("new/local/postgres/saveMonitorError").default = (...args) => {
  mockFn(`saveMonitorError`, args)
}
jest.mock(`new/local/utils/isProduction`, () => ({ default: jest.fn() }))
const { default: isProduction } = require("new/local/utils/isProduction")
const { default: HandleMonitorLog } = require("new/local/postgres/HandleMonitorLog")
describe(`new/local/postgres/HandleMonitorLog`, () => {
  afterEach(() => mockFn.mockClear())
  test.each([
    [`not do anything`, true, `error`],
    [`hide the message`, true, `notAnError`],
    [`save the message`, false, `error`],
    [`not do anything`, false, `notAnError`],
  ])(`should %s (isProduction %j, info.event %s)`, (str, isProductionValue, event) => {
    const info = { event }
    isProduction.mockReturnValueOnce(isProductionValue)
    HandleMonitorLog()(`msg`, info)
    expect({ info, mockFnCalls: mockFn.mock.calls }).toMatchSnapshot()
  })
})