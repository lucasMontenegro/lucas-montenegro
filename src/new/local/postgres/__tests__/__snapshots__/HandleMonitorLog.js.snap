// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`new/local/postgres/HandleMonitorLog should hide the message (isProduction true, info.event notAnError) 1`] = `
Object {
  "info": Object {
    "display": false,
    "event": "notAnError",
  },
  "mockFnCalls": Array [],
}
`;

exports[`new/local/postgres/HandleMonitorLog should not do anything (isProduction false, info.event notAnError) 1`] = `
Object {
  "info": Object {
    "event": "notAnError",
  },
  "mockFnCalls": Array [],
}
`;

exports[`new/local/postgres/HandleMonitorLog should not do anything (isProduction true, info.event error) 1`] = `
Object {
  "info": Object {
    "event": "error",
  },
  "mockFnCalls": Array [],
}
`;

exports[`new/local/postgres/HandleMonitorLog should save the message (isProduction false, info.event error) 1`] = `
Object {
  "info": Object {
    "event": "error",
  },
  "mockFnCalls": Array [
    Array [
      "saveMonitorError",
      Array [
        "msg",
        Object {
          "event": "error",
        },
      ],
    ],
  ],
}
`;
